core/internal/runhistoryreader/parquet/iterator/row_iterator.go

```diff
@@ -0,0 +1,391 @@
+package iterator
+
+// This file copies most of its logic from how history parquet files
+// are read on the backend.
+// See: https://github.com/wandb/core/blob/master/services/gorilla/internal/parquet/row_iterator.go
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"slices"
+
+	"github.com/apache/arrow-go/v18/parquet"
+	"github.com/apache/arrow-go/v18/parquet/file"
+	"github.com/apache/arrow-go/v18/parquet/metadata"
+	"github.com/apache/arrow-go/v18/parquet/pqarrow"
+	"github.com/apache/arrow-go/v18/parquet/schema"
+)
+
+const (
+	StepKey      = "_step"
+	TimestampKey = "_timestamp"
+)
+
+type HistoryPageParams struct {
+	MinStep int64 // Inclusive lower bound of _step
+	MaxStep int64 // Exclusive upper bound of _step
+}
+
+type EventsPageParams struct {
+	MinTimestamp float64
+	MaxTimestamp float64
+}
+
+// KeyValuePair is a key-value pair mapping,
+// it represents a single metric for a history row.
+// Where Key is the metric name and Value is the metric value.
+type KeyValuePair struct {
+	Key   string
+	Value any
+}
+
+// KeyValueList is a list of KeyValuePairs which represent a single history row.
+type KeyValueList []KeyValuePair
+
+// RowIterator is the primary interface to iterate over history rows
+// for a runs history parquet file.
+type RowIterator interface {
+	// Next advances the iterator to the next value.
+	Next() (bool, error)
+	// Value returns the current value of the iterator.
+	Value() KeyValueList
+	// Release releases the resources used by the iterator.
+	Release()
+}
+
+func selectColumns(schema *schema.Schema, columns []string) ([]int, error) {
+	if len(columns) == 0 {
+		indices := make([]int, schema.NumColumns())
+		for i := 0; i < len(indices); i++ {
+			indices[i] = i
+		}
+		return indices, nil
+	}
+
+	foundMap := map[string]struct{}{}
+
+	var indices []int
+	for i := 0; i < schema.NumColumns(); i++ {
+		col := schema.Column(i)
+		for _, selected := range columns {
+			if col.ColumnPath()[0] == selected {
+				indices = append(indices, i)
+				foundMap[selected] = struct{}{}
+			}
+		}
+	}
+
+	// Make sure that selected columns were found.
+	for _, selected := range columns {
+		if _, ok := foundMap[selected]; !ok {
+			return nil, fmt.Errorf("column %s selected but not found", selected)
+		}
+	}
+
+	return indices, nil
+}
+
+func selectRowGroups(pf *file.Reader, config IteratorConfig) ([]int, error) {
+	indices := []int{}
+
+	// filter row groups that are outside of the requested step range
+	metadata := pf.MetaData()
+	for i := 0; i < len(metadata.RowGroups); i++ {
+		minValue, maxValue, err := getColumnRangeFromStats(pf.MetaData(), i, config.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		if (config.Max == 0 || minValue < config.Max) && (config.Min == math.MaxInt64 || maxValue >= config.Min) {
```

Comment:
the exact comparisons are sketchy considering Max and Min are floats, see my other comment about it